### ISSUE TITLE
[Bug]: v0.3.1+ Single Download does not apply the correct `vid_type` when pulling from the ElasticSearch queue

### ISSUE THREAD

**Issue by lamusmaser** [tubearchivist/tubearchivist#412]
### I've read the documentation

- [X] I'm running the latest version of Tube Archivist and have read the [release notes](https://github.com/tubearchivist/tubearchivist/releases/latest).
- [X] I have read through the [wiki](https://github.com/tubearchivist/tubearchivist/wiki) and the [readme](https://github.com/tubearchivist/tubearchivist#installing-and-updating), particularly the [common errors](https://github.com/tubearchivist/tubearchivist#common-errors) section.

### Operating System

Unraid, Synology

## Describe the bug
When attempting to download a single video, the videos do not download due to a KeyError exception occurring. 

## Steps To Reproduce
Attempt to download a single video through the "Download Now" button on the Downloads Page.

## Expected behavior
Video should download instead of producing the exception.

### Relevant log output

```shell
erAoSrjDf6c: change status to priority
[2023-01-19 12:49:53,542: INFO/MainProcess] Task home.tasks.download_single[2078dc6a-9ee6-43a6-815d-168ff6bd203e] received
[2023-01-19 12:49:53,547: ERROR/ForkPoolWorker-4] Task home.tasks.download_single[2078dc6a-9ee6-43a6-815d-168ff6bd203e] raised unexpected: KeyError('vid_type')
Traceback (most recent call last):
  File "/root/.local/lib/python3.10/site-packages/celery/app/trace.py", line 451, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/root/.local/lib/python3.10/site-packages/celery/app/trace.py", line 734, in __protected_call__
    return self.run(*args, **kwargs)
  File "/app/home/tasks.py", line 117, in download_single
    "vid_type": pending_video["vid_type"],
KeyError: 'vid_type'
```

**Comment by bbilly1** [comment]
That should have gotten set during the migrations here: https://github.com/tubearchivist/tubearchivist/blob/3602ec45604a492d708cdf710302ade7f008aef4/tubearchivist/home/apps.py#L135-L160

That's for the videos in the download queue during migration, as well as your existing videos.

I'm guessing the startup checks aren't running or didn't run properly. Manually delete the leftover lock if needed:
```
docker exec -it archivist-redis redis-cli DEL ta:startup_check
```
then restart the container and monitor the logs for errors in during startup.

**Comment by lamusmaser** [comment]
This is specifically an issue with single downloads. Videos that are already downloaded are appropriately marked. Videos that are being added to the queue, however, have two different logic pieces - "priority" (or single downloads) was previously requesting their `vid_type` [here](https://github.com/tubearchivist/tubearchivist/pull/413/files#diff-9b5cf57a3f44d85a4722ad7d20fcd5b9db5cb8f6ada269ce44ca996e14ccd2e1L117); if they are in a queue, they are setting their `vid_type` to a default, like [here](url). #413 is expected to fix this issue by keeping the logic the same; an alternative is to make the response for both come from the source.

I definitely might have misinterpreted the calls, so we can work on the PR and get it requesting properly, but right now they do not perform the same type of call. Single Downloads are failing whereas queue downloads are processing properly.

**Comment by bbilly1** [comment]
OK, let's retrace this:
- Call comes in from the API here: https://github.com/tubearchivist/tubearchivist/blob/master/tubearchivist/api/views.py#L406-L426
- This will get the item as is from the pending list with `PendingInteract`: https://github.com/tubearchivist/tubearchivist/blob/3602ec45604a492d708cdf710302ade7f008aef4/tubearchivist/home/src/download/queue.py#L120-L124
- That video will have the `vid_type` set already. Old videos you had in the queue previously before updating will get a default `vid_type` of *videos* with the migrations as linked above https://github.com/tubearchivist/tubearchivist/issues/412#issuecomment-1399467568, new videos will get it during adding to the queue: https://github.com/tubearchivist/tubearchivist/blob/3602ec45604a492d708cdf710302ade7f008aef4/tubearchivist/home/src/download/queue.py#L269-L275
- You can verify the `vid_type` is set through the API, for example: `GET /api/download/<video-id>/`
- So when the task `download_single` gets called in the API view, it will have the same metadata as returned with `PendingInteract` which will have the `vid_type` key set.

If a video in the queue doesn't have the `vid_type` set, the migrations didn't run successfully. 

If you assume vid_type=videos like here https://github.com/tubearchivist/tubearchivist/pull/413/files#diff-9b5cf57a3f44d85a4722ad7d20fcd5b9db5cb8f6ada269ce44ca996e14ccd2e1R117 for all download now tasks, this won't later be able to identify streams and shorts, and they won't go to their correct category and channel page.

Does that make sense? Or is there something I'm missing?

**Comment by lamusmaser** [comment]
I believe the logic seems correct, and it should be testable, as these issues seem to have come through before we identified the lock issues. I should be able to add a video pre-v0.3.1, cut off the migrations during the upgrade to v0.3.2 (which would cause this issue to come through, if we are thinking it was a migrations issue), cause the error for single_download, then clear the startup_check lock only, restart the container, and that should fix the issue. If that works, then this is related to #414 and our work in that area should resolve the issue, which would mean the PR can be rejected.

Let me get that test going and get you the results.

**Comment by lamusmaser** [comment]
I was finally able to test this. Here is the sequence that I performed and the results.

TL;DR: `single_download` and downloading via the queue handle `vid_type` differently, we should see if we can align them. Based on my read here, the queue download is getting the `vid_type` actively from YouTube (or has an override for all to be `VideoTypeEnum.VIDEOS`) at time of download, whereas single_download is referencing the ES queue to get its `vid_type`.

### Steps to reproduce:
1. Install TA at v0.3.0
2. Setup a channel and add several videos to the queue (I set it to add 20 to the queue for ease of use).
3. Upgrade TA to v0.3.2
4. Immediately after, set the `startup_check` in `redis` so that it doesn't perform the check sequence, which doesn't allow the ES queue to update their internal `vid_type` reference.
5. Attempt a single_download - this will fail
6. Attempt to download the queue - this will succeed.

### Commands performed:
1. ```vi docker-compose.yml``` - Update version line to v0.3.0
2. ```./ytdl-test-repave.sh```
3. ```vi docker-compose.yml``` - Update version line to v0.3.2
4. ```./update-ytdl.sh && docker exec tubearchivist-test-redis redis-cli JSON.SET ta:startup_check . '{"status": true}'```

### The scripts perform the following:
## `ytdl-test-repave.sh`
1. Shut down and remove any active TA containers
2. Clear their associated local volumes (ES, Redis, Cache, and Media)
3. Pull the referenced image
4. Create and up the image, using forced-recreate and background daemon flags

## `update-ytdl.sh`
1. Pull the referenced image
2. Create and up the image, using the background daemon

### Relevant logs:
```
{
  "name" : "b60bf9dc2815",
  "cluster_name" : "docker-cluster",
  "cluster_uuid" : "VMr6EKZVSNewFoLLX6078A",
  "version" : {
    "number" : "8.6.0",
    "build_flavor" : "default",
    "build_type" : "docker",
    "build_hash" : "f67ef2df40237445caa70e2fef79471cc608d70d",
    "build_date" : "2023-01-04T09:35:21.782467981Z",
    "build_snapshot" : false,
    "lucene_version" : "9.4.2",
    "minimum_wire_compatibility_version" : "7.17.0",
    "minimum_index_compatibility_version" : "7.0.0"
  },
  "tagline" : "You Know, for Search"
}
sync redis
startup checks run in other thread
Migrations for 'home':
  home/migrations/0001_initial.py
    - Create model Account
sync redis
startup checks run in other thread
Operations to perform:
  Apply all migrations: admin, auth, authtoken, contenttypes, home, sessions
Running migrations:
  No migrations to apply.
[WARNING] This is not the first run! Skipping superuser creation.
To force it, remove /cache/initsu.lock
sync redis
startup checks run in other thread

231 static files copied to '/app/staticfiles', 971 post-processed.
[uWSGI] getting INI configuration from uwsgi.ini
*** Starting uWSGI 2.0.21 (64bit) on [Fri Jan 27 11:09:08 2023] ***
compiled with version: 10.2.1 20210110 on 14 January 2023 03:00:01
os: Linux-4.4.59+ #25556 SMP PREEMPT Sat Aug 28 02:17:26 CST 2021
detected binary path: /root/.local/bin/uwsgi
!!! no internal routing support, rebuild with pcre support !!!
*** Python threads support is disabled. You can enable it with --enable-threads ***
Python main interpreter initialized at 0x5596c377bc60
mapped 154032 bytes (150 KB) for 1 cores
*** Operational MODE: single process ***
sync redis
startup checks run in other thread
sync redis
sync redis
celery beat v5.2.7 (dawn-chorus) is starting.
WSGI app 0 (mountpoint='') ready in 2 seconds on interpreter 0x5596c377bc60 pid: 19 (default app)
uWSGI running as root, you can use --uid/--gid/--chroot options
spawned uWSGI worker 1 (pid: 28, cores: 1)
startup checks run in other thread
startup checks run in other thread
__    -    ... __   -        _
LocalTime -> 2023-01-27 11:09:11
Configuration ->
    . broker -> redis://tubearchivist-test-redis:6379//
    . loader -> celery.loaders.app.AppLoader
    . scheduler -> celery.beat.PersistentScheduler
    . db -> /cache/celerybeat-schedule
    . logfile -> [stderr]@%INFO
    . maxinterval -> 5.00 minutes (300s)
[2023-01-27 11:09:11,021: INFO/MainProcess] beat: Starting...
/root/.local/lib/python3.10/site-packages/celery/platforms.py:840: SecurityWarning: You're running the worker with superuser privileges: this is
absolutely not recommended!

Please specify a different user using the --uid option.


  warnings.warn(SecurityWarning(ROOT_DISCOURAGED.format(

 -------------- celery@f9569154d5f6 v5.2.7 (dawn-chorus)
  . home.tasks.run_manual_import
  . home.tasks.run_restore_backup
  . home.tasks.subscribe_to
  . run_backup
[2023-01-27 11:09:11,770: INFO/MainProcess] Connected to redis://tubearchivist-test-redis:6379//
[2023-01-27 11:09:11,776: INFO/MainProcess] mingle: searching for neighbors
[2023-01-27 11:09:12,795: INFO/MainProcess] mingle: all alone
[2023-01-27 11:09:12,829: INFO/MainProcess] celery@f9569154d5f6 ready.
yLNpy62jIFk: change status to priority
[2023-01-27 11:12:36,586: INFO/MainProcess] Task home.tasks.download_single[c8c54c92-3811-45f4-86b9-c5d2888a8057] received
[2023-01-27 11:12:36,593: ERROR/ForkPoolWorker-4] Task home.tasks.download_single[c8c54c92-3811-45f4-86b9-c5d2888a8057] raised unexpected: KeyError('vid_type')
Traceback (most recent call last):
  File "/root/.local/lib/python3.10/site-packages/celery/app/trace.py", line 451, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/root/.local/lib/python3.10/site-packages/celery/app/trace.py", line 734, in __protected_call__
    return self.run(*args, **kwargs)
  File "/app/home/tasks.py", line 117, in download_single
    "vid_type": pending_video["vid_type"],
KeyError: 'vid_type'
download pending
46ae1bf2-d8f1-485a-a543-53d9c1825e47: set task id
[2023-01-27 11:12:41,156: INFO/MainProcess] Task download_pending[46ae1bf2-d8f1-485a-a543-53d9c1825e47] received
[2023-01-27 11:12:41,382: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:12:44,367: WARNING/ForkPoolWorker-4] yLNpy62jIFk: get metadata from youtube
[2023-01-27 11:12:45,289: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:12:47,360: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:12:48,482: WARNING/ForkPoolWorker-4] l3zAqQacTGA: get metadata from youtube
[2023-01-27 11:12:49,456: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:12:49,715: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:12:51,071: WARNING/ForkPoolWorker-4] XI7Cxdj2pAQ: get metadata from youtube
[2023-01-27 11:12:52,108: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:12:52,320: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:12:53,382: WARNING/ForkPoolWorker-4] qIzfPsvRKMI: get metadata from youtube
[2023-01-27 11:12:54,214: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:12:54,423: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:12:55,487: WARNING/ForkPoolWorker-4] MVMCBsqi0L8: get metadata from youtube
[2023-01-27 11:12:56,437: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:12:56,804: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:12:57,867: WARNING/ForkPoolWorker-4] uQSM7070FMQ: get metadata from youtube
[2023-01-27 11:12:58,718: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:12:58,938: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:12:59,943: WARNING/ForkPoolWorker-4] wRQS5VGBGnM: get metadata from youtube
[2023-01-27 11:13:00,799: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:01,036: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:13:02,075: WARNING/ForkPoolWorker-4] r7THj30Fq_M: get metadata from youtube
[2023-01-27 11:13:02,995: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:03,310: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:13:04,479: WARNING/ForkPoolWorker-4] 5gYV2reIP-M: get metadata from youtube
[2023-01-27 11:13:05,347: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:05,559: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:13:06,686: WARNING/ForkPoolWorker-4] nynoN1lXUsA: get metadata from youtube
[2023-01-27 11:13:07,502: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:07,662: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:13:08,715: WARNING/ForkPoolWorker-4] BTZ5KVRUy1Q: get metadata from youtube
[2023-01-27 11:13:09,688: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:09,889: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:13:11,038: WARNING/ForkPoolWorker-4] nE0iffq9p0o: get metadata from youtube
[2023-01-27 11:13:11,846: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:12,146: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:13:13,189: WARNING/ForkPoolWorker-4] SR86aNrbWt0: get metadata from youtube
[2023-01-27 11:13:14,547: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:14,771: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:13:15,798: WARNING/ForkPoolWorker-4] sR6wfV_I4Ek: get metadata from youtube
[2023-01-27 11:13:16,695: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:16,977: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:13:18,048: WARNING/ForkPoolWorker-4] Soht0V1DvJA: get metadata from youtube
[2023-01-27 11:13:18,926: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:19,479: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:13:20,767: WARNING/ForkPoolWorker-4] sGu2f226qWQ: get metadata from youtube
[2023-01-27 11:13:22,027: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:22,469: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:13:23,461: WARNING/ForkPoolWorker-4] SwOgwUsLjEQ: get metadata from youtube
[2023-01-27 11:13:24,280: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:24,483: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:13:25,626: WARNING/ForkPoolWorker-4] mRfwKjlJKW4: get metadata from youtube
[2023-01-27 11:13:26,447: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:26,773: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:13:27,795: WARNING/ForkPoolWorker-4] xI4tBbz82nA: get metadata from youtube
[2023-01-27 11:13:28,691: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:28,941: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-01-27 11:13:30,135: WARNING/ForkPoolWorker-4] VAJiNvQskTA: get metadata from youtube
[2023-01-27 11:13:31,164: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-01-27 11:13:32,666: INFO/ForkPoolWorker-4] Task download_pending[46ae1bf2-d8f1-485a-a543-53d9c1825e47] succeeded in 51.5070168254897s: None
```

**Comment by lamusmaser** [comment]
Now, obviously, we could just handle the key issue and that would support this. But I still have a feeling we are getting something incorrectly for the queue download.

I want to do another check to see if I can confirm my suspicion.

**Comment by bbilly1** [comment]
OK, I think I see where the difference is. The add single video task gets the `vid_type` like this `"vid_type": pending_video["vid_type"]` here:

https://github.com/tubearchivist/tubearchivist/blob/master/tubearchivist/home/tasks.py#L115-L118

That will fail with the KeyError you see if that key isn't set.

On the other hand, when adding in bulk to the queue, this will use `get("vid_type", VideoTypeEnum.VIDEOS.value)` from the dict here:

https://github.com/tubearchivist/tubearchivist/blob/master/tubearchivist/home/src/download/yt_dlp_handler.py#L268-L273

So if the `vid_type` key isn't set, it will fallback to *videos* and continue.

That's a slight inconsistency, yes, as one will trigger an error and the other one will silently continue.

But the root problem of that are the migrations. If the migrations ran, the `vid_type` will always be set.

**Comment by lamusmaser** [comment]
Yeah, I was confirming the full aspect of the calls, but I think you've identified what I was seeing from my side. I think they should align, at least, to reduce the likelihood of a bug coming through in the future. I'm good with either option (`single_download` assumes it has been set and after v0.3.2, I would assume that we can safely make that assumption; queued downloads allows for a default if it doesn't find the reference, which is also good, but _could_ cause loss of information, but I don't think it is a significant enough issue to cause an override).

I think the PR itself will take action to keep them the same, so this creates consistency, but I'm good to implement (correctly this time) either option.

**Comment by lamusmaser** [comment]
As discussed on Discord, we are going to process with a default, similar to what is processed for the bulk download.

**Comment by lamusmaser** [comment]
Change the `single_download()` call to match the same as the bulk download call. Test on local network (same as before) finally allows for videos to download.

```
l3zAqQacTGA: change status to priority
[2023-02-05 11:27:57,303: INFO/MainProcess] Task home.tasks.download_single[6ce50e62-9c77-4fd1-b234-6c910ec37fc7] received
[2023-02-05 11:27:57,308: WARNING/ForkPoolWorker-4] Added to queue with priority: {'youtube_id': 'l3zAqQacTGA', 'vid_type': 'videos'}
[2023-02-05 11:27:57,444: WARNING/ForkPoolWorker-4] Downloading type: VideoTypeEnum.VIDEOS
[2023-02-05 11:27:59,248: WARNING/ForkPoolWorker-4] l3zAqQacTGA: get metadata from youtube
[2023-02-05 11:28:00,177: WARNING/ForkPoolWorker-4] UCsLiV4WJfkTEHH0b9PmRklw: get metadata from es
[2023-02-05 11:28:00,792: INFO/ForkPoolWorker-4] Task home.tasks.download_single[6ce50e62-9c77-4fd1-b234-6c910ec37fc7] succeeded in 3.4863181365653872s: None
```

**Comment by bbilly1** [comment]
Thanks for your help on this one. That's now merged in v0.3.3.
